### PR TITLE
Renderer state

### PR DESCRIPTION
**Pull request by brunoabinader**
Introduces a Memento-like RendererState object that will be responsible for fetching renderer state-related data (being camera positioning the first in the list), while preserving the encapsulation of `mbgl::UpdateParameters`.

This PR also fixes an inconsistency in our public APIs, in which `mbgl::CameraOptions` should only accept (and outputs) bearing and pitch values in degrees. We should only use such values in radians internally.

/cc @kkaefer @tmpsantos @1ec5 @tobrun

### CODE REVIEW

File: include/mbgl/renderer/renderer_state.hpp [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+#include <mbgl/map/camera.hpp>
+#include <mbgl/util/geo.hpp>
+
+namespace mbgl {
+
+class UpdateParameters;
+
+class RendererState {
+public:
+    static CameraOptions getCameraOptions(UpdateParameters&, const EdgeInsets& = {});
```
[the remark this answers is not in the document]
Comment (kkaefer) [reply include/mbgl/renderer/renderer_state.hpp]: This header could use some documentation clarifying what it does and how it can be used.

File: platform/android/src/native_map_view.cpp [hunk 2]
```diff
@@ -257,12 +257,12 @@ void NativeMapView::moveBy(jni::JNIEnv&, jni::jdouble dx, jni::jdouble dy, jni::
 void NativeMapView::jumpTo(jni::JNIEnv&, jni::jdouble angle, jni::jdouble latitude, jni::jdouble longitude, jni::jdouble pitch, jni::jdouble zoom) {
     mbgl::CameraOptions options;
     if (angle != -1) {
-        options.angle = -angle * util::DEG2RAD;
+        options.angle = angle;
```
Comment (kkaefer) [inline platform/android/src/native_map_view.cpp]: Was removing `-` your intention here (and in the other changesets in this file)

File: src/mbgl/map/transform.cpp [hunk 3]
```diff
@@ -462,7 +456,7 @@ void Transform::setAngle(double angle, const AnimationOptions& animation) {
 void Transform::setAngle(double angle, optional<ScreenCoordinate> anchor, const AnimationOptions& animation) {
     if (std::isnan(angle)) return;
     CameraOptions camera;
-    camera.angle = angle;
+    camera.angle = -angle * util::RAD2DEG;
```
Comment (kkaefer) [inline src/mbgl/map/transform.cpp]: Why did you move the negation to here?

File: include/mbgl/renderer/renderer_state.hpp [hunk 4]
```diff
@@ -0,0 +1,15 @@
+#pragma once
+
+#include <mbgl/map/camera.hpp>
+#include <mbgl/util/geo.hpp>
+
+namespace mbgl {
+
+class UpdateParameters;
+
+class RendererState {
```
Comment (kkaefer) [inline include/mbgl/renderer/renderer_state.hpp]: Why is this a class? It seems to only have one static function member. Could it be a standalone function instead?

File: platform/android/src/native_map_view.cpp (hunk 2 above)
Comment (brunoabinader) [reply platform/android/src/native_map_view.cpp]: Yes - that's only needed when converting bearing degrees into radians.

File: src/mbgl/map/transform.cpp (hunk 3 above)
Comment (brunoabinader) [reply src/mbgl/map/transform.cpp]: Same answer as above.

File: include/mbgl/renderer/renderer_state.hpp (hunk 4 above)
Comment (brunoabinader) [reply include/mbgl/renderer/renderer_state.hpp]: This class acts as a container for renderer state-like functions - I plan to add a few other letters, but  I'm :+1: in converting this into e.g. a sub-namespace instead.

File: include/mbgl/renderer/renderer_state.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (kkaefer) [reply include/mbgl/renderer/renderer_state.hpp]: What is a "Memento-like object"?

[the remark this answers is not in the document]
Comment (brunoabinader) [reply include/mbgl/renderer/renderer_state.hpp]: [Memento](https://en.wikipedia.org/wiki/Memento_pattern) is a design pattern that, among other things, exposes an internal state out of an opaque object, which is exactly what we needed to do here.

File: platform/android/src/native_map_view.cpp (hunk 2 above)
Comment (brunoabinader) [reply platform/android/src/native_map_view.cpp]: Specifically, the minus sign was being used to accommodate for the fact that the bearing is taken counterclockwise.

File: platform/ios/src/MGLMapView.mm [hunk 5]
```diff
@@ -1970,7 +1970,7 @@ - (MGLMapCamera *)cameraByRotatingToDirection:(CLLocationDirection)degrees aroun
     MGLMapCamera *camera;
     
     mbgl::ScreenCoordinate anchor = mbgl::ScreenCoordinate { anchorPoint.x, anchorPoint.y };
-    currentCameraOptions.angle = degrees * mbgl::util::DEG2RAD;
+    currentCameraOptions.angle = degrees;
```
Comment (1ec5) [inline platform/ios/src/MGLMapView.mm]: This incidentally fixes a missing `-`.

File: platform/darwin/src/MGLMapSnapshotter.mm [hunk 6]
```diff
@@ -550,7 +550,7 @@ - (void)setOptions:(MGLMapSnapshotOptions *)options
     if (CLLocationCoordinate2DIsValid(options.camera.centerCoordinate)) {
         cameraOptions.center = MGLLatLngFromLocationCoordinate2D(options.camera.centerCoordinate);
     }
-    cameraOptions.angle = MAX(0, options.camera.heading) * mbgl::util::DEG2RAD;
+    cameraOptions.angle = MAX(0, options.camera.heading);
```
Comment (1ec5) [inline platform/darwin/src/MGLMapSnapshotter.mm]: Before this PR, here was a bug where `MGLMapSnapshotOptions.camera`’s `heading` was misinterpreted to be counterclockwise instead of clockwise from north.

File: platform/darwin/src/MGLMapSnapshotter.mm [hunk 7]
```diff
@@ -550,7 +550,7 @@ - (void)setOptions:(MGLMapSnapshotOptions *)options
...
     cameraOptions.zoom = MAX(0, options.zoomLevel);
     cameraOptions.pitch = MAX(0, options.camera.pitch);
```
Comment (1ec5) [inline platform/darwin/src/MGLMapSnapshotter.mm]: Before this PR, here was a bug where `MGLMapSnapshotOptions.camera`’s `pitch` was misinterpreted to be in radians counterclockwise instead of degrees clockwise from north.

**Comment by brunoabinader** [comment]
> These accidentally fixed bugs should be noted in the iOS map SDK v4.5.0 and macOS map SDK v0.12.0 changelogs.

Glad this helped @1ec5 :tada:
